### ISSUE TITLE
Question about quantized values?

### ISSUE THREAD

**Issue by JingrenChen** [boluoweifenda/WAGE#6]
According to your paper, like figure5, the quantized values are always between (-1,+1) with uniform distance like 1/2, 1/4, 1/8, etc. So the storage of weights and computations are not exactly integers? (8-bit integers from -128 to +127). Or you mean scale them to int8 range, do the computation, then scale back to (-1,+1)? Maybe I misunderstood something, hope to find some answer here. Thanks.

**Comment by boluoweifenda** [comment]
@JingrenChen  We use integers to represent float from -1 to +1, it is a linear transformation, here is a example:
if value 1 -> 128, then 1/128 -> 1
              0.5 * 0.5           = 0.25
              64 *  64  / 128 =    32  (high 8 bits)
              0.5 + 0.25        = 0.75
              64  +   32         =   96

**Comment by JingrenChen** [comment]
Thanks for the quick reply, I have some further questions.
1. So the computations for entire net can actually be done in Int8 domain (-128 to +127), but in your code you simulate the effect of Int8 arithmetic in the corresponding linear transformation float (-1 to +1) right? 
2. According to figure5, are the weights of every layer always -0.5/0/+0.5? Or each layer has different scaling factor, therefore different ternary values?
3. Can this work be applied to normal pretrained models? For example, a pretrained MobileNet(fp32), then quantize to Int8 weights and activations for inference?

**Comment by boluoweifenda** [comment]
@JingrenChen 
1. Yes
2. Always -0.5/0/+0.5, but each layer has a scale factor, e.g. 32, then the equivalent weight value is -0.5/32   0    0.5/32
3. There are some implementations that have quantized the weights of mobilenet to INT8.
Our work mainly focus on backpropagation training with both integer operations and operands. If you only need inference and have floating-point arithmetic units, you can refer to the following method which has better accuracy:
[https://www.tensorflow.org/performance/quantization](url)

**Comment by JingrenChen** [comment]
It seems that Tensorflow's implementation still needs some floating-point arithmetic like dequantize, not entirely "integer-only". Maybe it's possible to extend your work to achieve similar accuracy like other 8-bit quantizations on ImageNet. Thanks for the discussion and Happy Dragon Boat Festival!

**Comment by boluoweifenda** [comment]
@JingrenChen 
端午安康 : )
